docs/Developers/NextGen_Core.rst

```diff
@@ -0,0 +1,334 @@
+  graph. When two objects (or nodes) are connected with a directed edge they
+  form a dependency relation. This lifts the limitation of GlobalBears being
+  dependent on LocalBears which was the case in the old core.
+
+The ``initialize_dependencies`` method in ``Core.py`` receives the bears that
+are to be run and processes bear dependencies using a consumer-based system so
+that each dependency bear has only one instance per section and file-dict. It
+returns a set of dependency bears along with those bears that don't have any
+dependencies or whose dependencies have been resolved (these are the ones that
+are scheduled to be run). The bears are resolved by traversing each edge of the
+dependency graph and instantiating dependency bears accordingly one per section.
+Before the bears are run we initialize the dependency tracking in the
+``__init__`` method of the class ``Session`` which is responsible for running
+coala sessions. The ``dependency_tracker`` object is used while scheduling and
+cleaning up bears and also while finishing bear tasks.
+
+The bears that have no dependencies or whose dependencies have been resolved,
+only their tasks will be scheduled for execution. Before executing any task
+coala looks it up in the cache. In case of a hit, the existing results that are
+stored in the cache for the corresponding task arguments are called using
+``execute_task_with_cache`` method. In case of a miss or if coala is run without
+a cache the task is executed. The bears without any running tasks are cleaned up
+from the state of an ongoing run by resolving its dependencies, scheduling
+dependent bears and removing the bear from the ``running_tasks`` dict. When a
+task belonging to a particular bear completes, then ``_finish_task`` method is
+responsible for checking whether the bear execution is complete and the result
+generated from the task have been returned. If any new tasks are to be scheduled
+then this method is responsible for doing so after the dependencies have been
+resolved.
+
+Even though bears still have to pass ``Result`` instances to communicate with
+coala, it is now possible to pass arbitrary Python objects. Dependency bears
+benefit from this because now they can pass data according to their needs
+without being bound to only ``Result`` objects.
+
+The ``dependency_results`` lie inside ``self.dependecy_results`` inside a bear
+and can be accessed that way. **Although this is highly discouraged since it
+bypasses caching and could yield weird results when the core is run multiple
+times in a row.**
+
+``DependencyBear``
+------------------
+
+Handling of bear dependencies by the old core wasn't effective. The
+old core used a queuing mechanism to communicate between bear runs and passed
+all the dependencies to the ``dependency_results`` argument of the ``run``
+function which led to problems of a ``GlobalBear`` not running if any one of its
+dependencies (e.g. a ``LocalBear``) terminated. The NextGen-Core improves on
```

Comment:
actually, if a dependency terminates, the dependent component gets unscheduled too^^ This is desired^^